### PR TITLE
Attempt to sum up testing discussion.

### LINKED ISSUE #61: Testing framework (web)
Pulling this out of https://github.com/red-gate/Tech-Radar/issues/27:

> JavaScript and, more also specifically, React testing is definitely an area we haven't pinned down.

> We've been using jest for general js testing, which appears to work just fine.
> We've successfully written some React component tests with jest and enzyme but haven't found them particularly useful beyond a proof of concept

we have a few different testing stacks so far:

- for running unit tests outside of the browser:
  - jest (with optional enzyme for testing react components)
- for running unit tests inside the browser:
  - karma+jasmine
- for driving an entire web application:
  - figleaf is currently using jest+spectron (which is chromedriver under the hood) to run a basic test against an installed product
  - it looks like monitor has some basic nunit+selenium (which is also chromedriver) smoke tests

### PR DESCRIPTION
**Pull request by fffej**
Resolves #61 

- [x] Are you happy for the content of this change to be publicly visible?
- [x] Does `radar.csv` render correctly when viewed on Github?
- [x] Does the [updated tech radar](https://radar.thoughtworks.com/?sheetId=https%3A%2F%2Fraw.githubusercontent.com%2Fred-gate%2FTech-Radar%2Fweb-test%2Fradar.csv) display as you expect?

### CODE REVIEW

File: radar.csv [hunk 1]
```diff
@@ -108,7 +108,14 @@ Physical whiteboards,adopt,tools,true,"Some teams use physical whiteboards for i
...
-"Zero bugs policy",explore,techniques,true,"A <a href='https://medium.com/qualityfaster/the-zero-bug-policy-b0bd987be684'>zero-bug policy</a> does not mean bug-free code production; it means striving to eradicate all known bugs. ICloneable<Sheep> will be exploring this with support from the Quality Coaches."
+"Zero bugs policy",explore,techniques,true,"A <a href='https://medium.com/qualityfaster/the-zero-bug-policy-b0bd987be684'>zero-bug policy</a> does not mean bug-free code production; it means striving to eradicate all known bugs. ICloneable&lt;Sheep&gt; will be exploring this with support from the Quality Coaches."
+"Karma",adopt,languages and frameworks,true,"<a href='https://karma-runner.github.io/latest/index.html'>Karma</a> is a test runner for JavaScript which spawns a web server that executes source code against test code for each of the browsers connected."
+"jasmine",explore,languages and frameworks,true,"<a href='https://jasmine.github.io/'>Jasmine</a> is a BDD framework for testing JavaScript code. The Figleaf team is evaluating this."
```
Comment (nyctef) [inline radar.csv]: Sorry, realized the way I wrote this in the issue was a bit misleading - Figleaf isn't currently using jasmine, but Monitor are using it through Karma.

File: radar.csv [hunk 2]
```diff
@@ -108,8 +108,16 @@ Physical whiteboards,adopt,tools,true,"Some teams use physical whiteboards for i
...
+"jasmine",explore,languages and frameworks,true,"<a href='https://jasmine.github.io/'>Jasmine</a> is a BDD framework for testing JavaScript code. The Monitor team are using it through Karma."
+"nightwatch",explore,languages and frameworks,true,"<a href='http://nightwatchjs.org/'>Nightwatch</a> is an end-to-end testing solution for browser-based apps using WebDriver. The ICloneable&lt;Sheep&gt; team is exploring this."
+"jest",adopt,languages and frameworks,true,"<a href='https://jestjs.io/'>Jest</a> is a way to test all JavaScript code that's used by Facebook. COREDev have a non-trivial amount of tests over multiple projects that use jest."
+"spectron",explore,languages and frameworks,true,"<a href='https://github.com/electron/spectron'>spectron</a> is a tool to allow you to easily test your Electron apps using ChromeDriver/WebDrive IO. Figleaf is exploring this."
+"mocha",explore,languages and frameworks,true,"<a href='https://mochajs.org'>mocha</a> is a feature rich test framework running on Node/Browser. The ICloneable&lt;Sheep&gt; team are exploring this."
```
Comment (Praveen-Redgate) [inline radar.csv]: In IClonableSheep we are currently using `mocha` and `chai`, we are exploring `jest`.

**Comment by ChrisHurley** [comment]
Just to be clear about that, we are not exploring `mocha` and `chai` as something new - we've been using them for ages and not that heavily. It might make more sense for us to migrate onto `jest`.

File: radar.csv (hunk 2 above)
Comment (fffej) [reply radar.csv]: Hmm. Should I put `mocha` and `chai` in adopt? (as in your confident that it's a great default choice) or should I leave them in `explore` as it's possible you might choose `jest` to be in adopt?

Comment (Alex-Kondratov-Redgate) [reply radar.csv]: I agree. From my experience `jest` is a good all purpose testing framework, so it doesn't make sense to keep `mocha `and `chai` in addition to it if we don't have any strong reasons for that.

**Comment by garethbragg** [comment]
@ChrisHurley Would it be too strong to put `mocha` and `chai` in _Endure_ ("It is OK for this to be used in Product Development, but don't add more of it and migrate off it wherever possible")?

**Comment by ChrisHurley** [comment]
I'd be fine with either Endure or Retire, to be honest. We don't have a big investment in `mocha`/`chai` at the moment (101 tests).

**Comment by fffej** [comment]
I've updated to Endure (simply because Mocha/Chai are well used in CORE Dev). Putting in retire seems wrong.

**Comment by deneboulton** [comment]
> I've updated to Endure (simply because Mocha/Chai are well used in CORE Dev). Putting in retire seems wrong at the moment.

We have used mocha but we have moved away from it already (same with Jasmine & Karma), actively started doing that about a year ago. 

Active projects in this area make use of jest with enzyme, we are happy with testing using the default jsdom approach that jest offers.

Regarding chai, which is a library for assertions, we do use chai-enzyme, which naturally depends upon chai. This offers chai assertion language for when writing enzyme tests, I used this because the out of the box enzyme selectors can be quite cumbersome.

I am not totally sold on chai-enzyme because it means our test suites sometimes make use of jest and chai-enzyme assertions in the the same suite, depending on what is being tested, which can read a bit weird to new maintainers.

We have some legacy JS tests for the Website that we never touch, and haven't in years, use `Chutzpah` as a test runner. RETIRE :)

**Comment by deneboulton** [comment]
> I've updated to Endure (simply because Mocha/Chai are well used in CORE Dev). Putting in retire seems wrong at the moment.

We have used mocha but we have moved away from it already (same with Jasmine & Karma), actively started doing that about a year ago. 

Active projects in this area make use of jest with enzyme, we are happy with testing using the default jsdom approach that jest offers.

Regarding chai, which is a library for assertions, we do use chai-enzyme, which naturally depends upon chai. This offers chai assertion language for when writing enzyme tests, I used this because the out of the box enzyme selectors can be quite cumbersome.

I am not totally sold on chai-enzyme because it means our test suites sometimes make use of jest and chai-enzyme assertions in the the same suite, depending on what is being tested, which can read a bit weird to new maintainers.

We have some legacy JS tests for the Website that we never touch, and haven't in years, use `Chutzpah` as a test runner. RETIRE :)

**Comment by fffej** [comment]
OK, last call here.

I've moved `mocha` to Retire since CORE seem to be doing that already (@deneboulton)  and @ChrisHurley didn't have any objections 😄.

If you're happy with this, press Merge.

**Comment by deneboulton** [comment]
🎆

**Pull request merged**